### PR TITLE
santad: Wait for driver appearance using IOKit notifications.

### PR DESCRIPTION
**Pull request by russellhancox**

### CODE REVIEW

File: Source/santad/SNTDriverManager.m [hunk 1]
```diff
@@ -86,6 +70,61 @@ - (void)dealloc {
...
+    dispatch_semaphore_signal(sema);
+  };
+
+  LOGI(@"Waiting for Santa driver to become available");
+
+  IOServiceAddMatchingNotification(notificationPort,
+                                   kIOMatchedNotification,
+                                   matchingDict,
+                                   driverAppearedHandler,
+                                   (__bridge_retained void *)block,
```
Comment (tburgin) [inline Source/santad/SNTDriverManager.m]: Does the block fire on a different thread? Just want to make sure the below dispatch wait won’t block the execution of the block due to notification.

Comment (russellhancox) [reply Source/santad/SNTDriverManager.m]: It did, I've changed the scope of the PR.

File: Source/common/SNTXPCControlInterface.h [hunk 2]
```diff
@@ -37,6 +37,7 @@
 - (void)flushCache:(void (^)(BOOL))reply;
 - (void)cacheBucketCount:(void (^)(NSArray *))reply;
 - (void)checkCacheForVnodeID:(santa_vnode_id_t)vnodeID withReply:(void (^)(santa_action_t))reply;
+- (void)connectionEstablished:(void (^)(BOOL))reply;
```
Comment (tburgin) [inline Source/common/SNTXPCControlInterface.h]: What connection? The driver?

File: Source/santactl/Commands/SNTCommandStatus.m [hunk 3]
```diff
@@ -190,6 +196,7 @@ - (void)runWithArguments:(NSArray *)arguments {
     printf("%s\n", [statsStr UTF8String]);
   } else {
     printf(">>> Daemon Info\n");
+    printf("  %-25s | %s\n", "Connected", driverConnected ? "Yes" : "No");
```
Comment (tburgin) [inline Source/santactl/Commands/SNTCommandStatus.m]: Something like: "Driver Connected"?
Also, add this to the JSON output.

File: Source/santad/SNTDriverManager.m [hunk 4]
```diff
@@ -23,67 +23,97 @@
...
+}
 
-      if (!serviceObject) {
-        LOGD(@"Waiting for Santa driver to become available");
-        sleep(delay);
-        if (delay < MAX_DELAY) delay *= 2;
-      }
-    } while (!serviceObject);
-    CFRelease(classToMatch);
+- (void)dealloc {
```
Comment (tburgin) [inline Source/santad/SNTDriverManager.m]: Given SNTDriverManager is essentially a singleton, this should never be called. Add a note.

File: Source/santad/SNTDriverManager.m [hunk 5]
```diff
@@ -23,67 +23,97 @@
...
+// Helper function used with IOServiceAddMatchingNotification which expects
+// a DriverAppearedBlock to be passed as the 'reference' argument. The block
+// will be called for each object iterated over and if the block wants to keep
+// the object, it should call IOObjectRetain().
+typedef void (^DriverAppearedBlock)(io_object_t object);
+static void driverAppearedHandler(void *info, io_iterator_t iterator) {
+  DriverAppearedBlock block = (__bridge DriverAppearedBlock)info;
+  io_object_t object = 0;
+  while ((object = IOIteratorNext(iterator))) {
+    block(object);
```
Comment (tburgin) [inline Source/santad/SNTDriverManager.m]: You may want to double check block() does indeed exist.

File: Source/santad/SNTDriverManager.m [hunk 6]
```diff
@@ -23,67 +23,97 @@
...
+  DriverAppearedBlock block = ^(io_object_t object) {
     // This calls `initWithTask`, `attach` and `start` in `SantaDriverClient`
-    kr = IOServiceOpen(serviceObject, mach_task_self(), 0, &_connection);
-    IOObjectRelease(serviceObject);
+    kern_return_t kr = IOServiceOpen(object, mach_task_self(), 0, &_connection);
     if (kr != kIOReturnSuccess) {
-      LOGD(@"Failed to open Santa driver service");
-      return nil;
+      LOGE(@"Failed to open santa-driver service: 0x%X", kr);
+      exit(1);
```
Comment (tburgin) [inline Source/santad/SNTDriverManager.m]: Is exiting the best option for all these cases? I could see santad flapping.

Comment (russellhancox) [reply Source/santad/SNTDriverManager.m]: I can't see a better option - if establishing a connection failed after the driver appeared there's no point waiting for the driver to appear again and there's not a notification that whatever caused connection failures is now fixed. Exiting and hoping that the next time starting might lead to a better outcome is about all we can do.

File: Source/santad/SNTDriverManager.m (hunk 5 above)
Comment (russellhancox) [reply Source/santad/SNTDriverManager.m]: Done

File: Source/santad/SNTDriverManager.m (hunk 4 above)
Comment (russellhancox) [reply Source/santad/SNTDriverManager.m]: We might treat SNTDriverManager like a singleton but it isn't implemented as one.

File: Source/santactl/Commands/SNTCommandStatus.m (hunk 3 above)
Comment (russellhancox) [reply Source/santactl/Commands/SNTCommandStatus.m]: Done

File: Source/common/SNTXPCControlInterface.h (hunk 2 above)
Comment (russellhancox) [reply Source/common/SNTXPCControlInterface.h]: I've renamed to make this clearer

**Comment by tburgin** [comment]
LGTM

**Pull request merged**